Android: Add support for cursor icons

thebestnom opened the pull request:
Adds support for cursor shapes on Android

Custom cursor shape is way harder, there need to be a way to move image from native to java by jni and I can't seem to find any way to do it, any help would be appriciated

I'll add a branch for 3.2 with https://github.com/godotengine/godot/pull/43249 together once this is ready

@m4gr3d

thebestnom committed:
Android: Add support for cursor icons

thebestnom commented:
Anythinhg else on this? @m4gr3d

m4gr3d reviewed (changes requested):
The PR looks good. I've added a few pieces of feedback.

On the java side, the same logic is repeated both in the `GodotGLRenderView` and `GodotVulkanRenderView`. Move that logic into a helper class (e.g: `GodotPointerIconHandler`), and have both instances of `GodotRenderView` forward the necessary calls to that class instead.

Let's move support for custom cursor shape into another PR. From the look of the Android api, only passing [`Bitmap`](https://developer.android.com/reference/android/graphics/Bitmap) is supported for custom pointer icon. The Godot native api has a matching [`Bitmap`](https://github.com/godotengine/godot/blob/master/scene/resources/bit_map.h) resource class so we may be able to add a conversion utility between the two (and any other matching Godot resource types).

thebestnom commented:
> Let's move support for custom cursor shape into another PR. From the look of the Android api, only passing Bitmap is supported for custom pointer icon. The Godot native api has a matching Bitmap resource class so we may be able to add a conversion utility between the two (and any other matching Godot resource types).

Yeah, that what I was going to do

I already tried a way to convert native bitmap to java bitmap but jni didn't like that, the best way I can think of is sending bitmap as array, but Ill think of a way when well get there

For now Ill try to fix the code this week, thanks for the cr ☺️

thebestnom commented on platform/android/display_server_android.cpp:
```diff
@@ -871,3 +874,40 @@ int DisplayServerAndroid::_android_button_mask_to_godot_button_mask(int android_
+		1015,
+		1014,
+		1003,
+	};
```
Ok, added comment from what it's godot wise, do you want me to comment full map from godot to android?

m4gr3d commented on platform/android/display_server_android.h:
```diff
@@ -70,6 +70,28 @@ class DisplayServerAndroid : public DisplayServer {
 
 	int buttons_state;
 
+	// https://developer.android.com/reference/android/view/PointerIcon
+	// couldn't found a way to import that from ndk
```
Can you update this documentation. Should be something like `Mapping from Android PointerIcon types to CursorShape values`.

thebestnom commented:
@m4gr3d tried fixing the comment, while doing this I rebased buy it looks like android build stop working, trying to fix

thebestnom commented:
looks like it's broken on main

thebestnom commented:
Ok, it's crashing on `rendering_device_vulkan.cpp` line 5477 with a `Signal: SIGSEGV (signal SIGSEGV: invalid address (fault address: 0x18))`...
I'm not good enough with vulkan (or 3d api in general 😅) but Ill try to fix it

akien-mga commented:
Code looks good to me, is this ready to merge @thebestnom @m4gr3d?

thebestnom commented:
Last time I tested that it worked, but I can't run master on phone for about 4 month (can't remember exactly) so I haven't tested that since

thebestnom commented:
@akien-mga ok, finally could test today and looks like it is still working!

m4gr3d commented:
> @akien-mga ok, finally could test today and looks like it is still working!

Great! Are all the requested changes addressed?

thebestnom commented:
> > @akien-mga ok, finally could test today and looks like it is still working!
> 
> Great! Are all the requested changes addressed?

Looked at the pr comments and the changes to refresh my memories and looks like I already addressed the comments, yep

m4gr3d commented on platform/android/java_godot_view_wrapper.cpp:
```diff
@@ -64,6 +65,11 @@ void GodotJavaViewWrapper::release_pointer_capture() {
 }
 
+void GodotJavaViewWrapper::set_pointer_icon(int pointer_type) {
+	JNIEnv *env = get_jni_env();
+	env->CallVoidMethod(_godot_view, _set_pointer_icon, pointer_type);
```
Follow the pattern in the preceding methods:
1. Check that `_set_pointer_icon` is valid.
2. Check that `env` is not `null`.

thebestnom replied:
fixed 😄

thebestnom commented:
ok, finally fixed static checks... I think this is ready for merge

akien-mga commented:
Thanks!
